src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -68,125 +70,153 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
         /// </summary>
         internal TStructuralType ParentObject { get; set; }
 
+        /// <summary>
+        /// API Handler Factory.
+        /// </summary>
+        internal ODataAPIHandlerFactory ApiHandlerFactory { get; set; }
```

Comment:
Avoid internal set property in a public virtual method.

We can pass the "Factory" to the lower layer api using argument.